From a9cef1ff33d1ae7cfaa7bf78d02a4d44bdf2983c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Johannes=20M=C3=BCller?= <straightshoota@gmail.com>
Date: Wed, 20 May 2026 11:34:02 +0200
Subject: [PATCH] Add support for OpenSSL 4.0 (#16839)

[OpenSSL 4.0.0] is out and this brings compatibility with the new release, following the [migration guide].
Still waiting for package availability in order to install `openssl` in CI workflow.
Alpine Linux seems to be in no hurry to add support for OpenSSL 4 (they don't even have 3.6).

[OpenSSL 4.0.0]: https://github.com/openssl/openssl/releases/tag/openssl-4.0.0
[migration guide]: https://docs.openssl.org/4.0/man7/ossl-guide-migration/#openssl-40

Co-authored-by: Julien Portalier <julien@portalier.com>
---
 spec/std/openssl/ssl/context_spec.cr   |  8 --------
 src/openssl/lib_crypto.cr              |  9 ++++++---
 src/openssl/lib_ssl.cr                 | 14 ++++++++------
 src/openssl/ssl/hostname_validation.cr |  6 +++---
 src/openssl/x509/name.cr               |  2 +-
 5 files changed, 18 insertions(+), 21 deletions(-)

diff --git a/spec/std/openssl/ssl/context_spec.cr b/spec/std/openssl/ssl/context_spec.cr
index 02ef634fef6e..56f330dafe46 100644
--- a/spec/std/openssl/ssl/context_spec.cr
+++ b/spec/std/openssl/ssl/context_spec.cr
@@ -19,8 +19,6 @@ describe OpenSSL::SSL::Context do
 
     context.modes.should eq(OpenSSL::SSL::Modes.flags(AUTO_RETRY, RELEASE_BUFFERS))
     context.verify_mode.should eq(OpenSSL::SSL::VerifyMode::PEER)
-
-    OpenSSL::SSL::Context::Client.new(LibSSL.tlsv1_method)
   end
 
   it "new for server" do
@@ -36,8 +34,6 @@ describe OpenSSL::SSL::Context do
 
     context.modes.should eq(OpenSSL::SSL::Modes.flags(AUTO_RETRY, RELEASE_BUFFERS))
     context.verify_mode.should eq(OpenSSL::SSL::VerifyMode::NONE)
-
-    OpenSSL::SSL::Context::Server.new(LibSSL.tlsv1_method)
   end
 
   it "insecure for client" do
@@ -46,8 +42,6 @@ describe OpenSSL::SSL::Context do
     context.verify_mode.should eq(OpenSSL::SSL::VerifyMode::NONE)
     context.options.no_ssl_v3?.should_not be_true
     context.modes.should eq(OpenSSL::SSL::Modes::AUTO_RETRY)
-
-    OpenSSL::SSL::Context::Client.insecure(LibSSL.tlsv1_method)
   end
 
   it "insecure for server" do
@@ -56,8 +50,6 @@ describe OpenSSL::SSL::Context do
     context.verify_mode.should eq(OpenSSL::SSL::VerifyMode::NONE)
     context.options.no_ssl_v3?.should_not be_true
     context.modes.should eq(OpenSSL::SSL::Modes::AUTO_RETRY)
-
-    OpenSSL::SSL::Context::Server.insecure(LibSSL.tlsv1_method)
   end
 
   it "sets certificate chain" do
diff --git a/src/openssl/lib_crypto.cr b/src/openssl/lib_crypto.cr
index 007de302d2ba..d410d0ed65e4 100644
--- a/src/openssl/lib_crypto.cr
+++ b/src/openssl/lib_crypto.cr
@@ -1,6 +1,6 @@
 # Supported library versions:
 #
-# * openssl (1.1.0–3.3+)
+# * openssl (1.1.0–4.0+)
 # * libressl (2.0–4.0+)
 #
 # See https://crystal-lang.org/reference/man/required_libraries.html#tls
@@ -24,7 +24,7 @@
       {% from_libressl = (`sh -c 'hash pkg-config 2> /dev/null || printf %s false'` != "false") &&
                          (`sh -c 'test -f $(pkg-config --silence-errors --variable=includedir libcrypto)/openssl/opensslv.h || printf %s false'` != "false") &&
                          (`sh -c 'printf "#include <openssl/opensslv.h>\nLIBRESSL_VERSION_NUMBER" | ${CC:-cc} $(pkg-config --cflags --silence-errors libcrypto || true) -E -'`.chomp.split('\n').last != "LIBRESSL_VERSION_NUMBER") %}
-      {% ssl_version = `sh -c 'hash pkg-config 2> /dev/null && pkg-config --silence-errors --modversion libcrypto || printf %s 0.0.0'`.split.last.gsub(/[^0-9.]/, "") %}
+      {% ssl_version = `sh -c 'hash pkg-config 2> /dev/null && pkg-config --silence-errors --modversion libcrypto || printf %s 0.0.0'`.split.last.gsub(/[^0-9.]/, "").gsub(/\.0(\d)/, ".\\1") %}
     {% end %}
 
     {% if from_libressl %}
@@ -177,7 +177,10 @@ lib LibCrypto
   fun obj_find_sigid_algs = OBJ_find_sigid_algs(sigid : Int32, pdig_nid : Int32*, ppkey_nid : Int32*) : Int32
 
   fun asn1_object_free = ASN1_OBJECT_free(obj : ASN1_OBJECT)
-  fun asn1_string_data = ASN1_STRING_data(x : ASN1_STRING) : Char*
+  {% if compare_versions(OPENSSL_VERSION, "4.0.0") < 0 %}
+    fun asn1_string_data = ASN1_STRING_data(x : ASN1_STRING) : Char*
+  {% end %}
+  fun asn1_string_get0_data = ASN1_STRING_get0_data(x : ASN1_STRING) : Char*
   fun asn1_string_length = ASN1_STRING_length(x : ASN1_STRING) : Int
   fun asn1_string_print = ASN1_STRING_print(out : Bio*, v : ASN1_STRING) : Int
   fun i2t_asn1_object = i2t_ASN1_OBJECT(buf : Char*, buf_len : Int, a : ASN1_OBJECT) : Int
diff --git a/src/openssl/lib_ssl.cr b/src/openssl/lib_ssl.cr
index 980c14c0f5b9..fa6b8bfd03a7 100644
--- a/src/openssl/lib_ssl.cr
+++ b/src/openssl/lib_ssl.cr
@@ -6,8 +6,8 @@ require "./lib_crypto"
 
 # Supported library versions:
 #
-# * openssl (1.1.0–3.3+)
-# * libressl (2.0–4.0+)
+# * openssl (1.1.0–4.0+)
+# * libressl (3.0–4.0+)
 #
 # See https://crystal-lang.org/reference/man/required_libraries.html#tls
 {% begin %}
@@ -30,7 +30,7 @@ require "./lib_crypto"
       {% from_libressl = (`sh -c 'hash pkg-config 2> /dev/null || printf %s false'` != "false") &&
                          (`sh -c 'test -f $(pkg-config --silence-errors --variable=includedir libssl)/openssl/opensslv.h || printf %s false'` != "false") &&
                          (`sh -c 'printf "#include <openssl/opensslv.h>\nLIBRESSL_VERSION_NUMBER" | ${CC:-cc} $(pkg-config --cflags --silence-errors libssl || true) -E -'`.chomp.split('\n').last != "LIBRESSL_VERSION_NUMBER") %}
-      {% ssl_version = `sh -c 'hash pkg-config 2> /dev/null && pkg-config --silence-errors --modversion libssl || printf %s 0.0.0'`.split.last.gsub(/[^0-9.]/, "") %}
+      {% ssl_version = `sh -c 'hash pkg-config 2> /dev/null && pkg-config --silence-errors --modversion libssl || printf %s 0.0.0'`.split.last.gsub(/[^0-9.]/, "").gsub(/\.0(\d)/, ".\\1") %}
     {% end %}
 
     {% if from_libressl %}
@@ -208,9 +208,11 @@ lib LibSSL
   # TLS alert codes
   SSL_AD_INTERNAL_ERROR = 80
 
-  fun tlsv1_method = TLSv1_method : SSLMethod
-  fun tlsv1_1_method = TLSv1_1_method : SSLMethod
-  fun tlsv1_2_method = TLSv1_2_method : SSLMethod
+  {% if compare_versions(OPENSSL_VERSION, "4.0.0") < 0 %}
+    fun tlsv1_method = TLSv1_method : SSLMethod
+    fun tlsv1_1_method = TLSv1_1_method : SSLMethod
+    fun tlsv1_2_method = TLSv1_2_method : SSLMethod
+  {% end %}
 
   fun ssl_get_error = SSL_get_error(handle : SSL, ret : Int) : SSLError
   fun ssl_get_servername = SSL_get_servername(ssl : SSL, host_type : TLSExt) : UInt8*
diff --git a/src/openssl/ssl/hostname_validation.cr b/src/openssl/ssl/hostname_validation.cr
index 8c0a787e6541..8d874918df99 100644
--- a/src/openssl/ssl/hostname_validation.cr
+++ b/src/openssl/ssl/hostname_validation.cr
@@ -37,14 +37,14 @@ module OpenSSL::SSL::HostnameValidation
 
         case current_name.type
         when LibCrypto::GEN_DNS
-          dns_name = LibCrypto.asn1_string_data(current_name.value)
+          dns_name = LibCrypto.asn1_string_get0_data(current_name.value)
           dns_name_len = LibCrypto.asn1_string_length(current_name.value)
           return Result::MalformedCertificate if dns_name_len != LibC.strlen(dns_name)
 
           pattern = String.new(dns_name, dns_name_len)
           return Result::MatchFound if matches_hostname?(pattern, hostname)
         when LibCrypto::GEN_IPADD
-          data = Slice.new(LibCrypto.asn1_string_data(current_name.value), LibCrypto.asn1_string_length(current_name.value))
+          data = Slice.new(LibCrypto.asn1_string_get0_data(current_name.value), LibCrypto.asn1_string_length(current_name.value))
 
           case data.size
           when 4
@@ -90,7 +90,7 @@ module OpenSSL::SSL::HostnameValidation
     asn1 = LibCrypto.x509_name_entry_get_data(name_entry)
     return Result::Error if asn1.null?
 
-    str = LibCrypto.asn1_string_data(asn1)
+    str = LibCrypto.asn1_string_get0_data(asn1)
     str_len = LibCrypto.asn1_string_length(asn1)
     return Result::MalformedCertificate if str_len != LibC.strlen(str)
 
diff --git a/src/openssl/x509/name.cr b/src/openssl/x509/name.cr
index dacb04d67c2c..4a1bc97fe133 100644
--- a/src/openssl/x509/name.cr
+++ b/src/openssl/x509/name.cr
@@ -86,7 +86,7 @@ module OpenSSL::X509
         end
 
         asn1 = LibCrypto.x509_name_entry_get_data(entry)
-        str = LibCrypto.asn1_string_data(asn1)
+        str = LibCrypto.asn1_string_get0_data(asn1)
         str_len = LibCrypto.asn1_string_length(asn1)
 
         {oid, String.new(str, str_len)}
